src/vm/callables.rs

```diff
@@ -0,0 +1,84 @@
+use vm::errors::{InterpreterResult as Result, Error};
+use vm::representations::SymbolicExpression;
+use vm::types::TypeSignature;
+use vm::{eval, Value, Context, Environment};
+
+pub enum CallableType <'a> {
+    UserFunction(DefinedFunction),
+    NativeFunction(&'a Fn(&[Value]) -> Result<Value>),
+    SpecialFunction(&'a Fn(&[SymbolicExpression], &mut Environment, &Context) -> Result<Value>)
+}
+
+#[derive(Clone)]
+pub struct DefinedFunction {
+    is_public: bool,
+    types: Option<Vec<TypeSignature>>,
+    pub arguments: Vec<String>,
+    pub body: SymbolicExpression
+}
+
+#[derive(Clone,PartialEq,Eq,Hash)]
+pub struct FunctionIdentifier {
+    pub arguments: Vec<String>,
+    pub body: SymbolicExpression
+}
+
+
+impl DefinedFunction {
+    pub fn apply(&self, args: &[Value], env: &mut Environment) -> Result<Value> {
+        let mut context = Context::new();
+
+        if !self.is_public {
+            let arg_iterator = self.arguments.iter().zip(args.iter());
+            for (arg, value) in arg_iterator {
+                if let Some(_) = context.variables.insert(arg.clone(), value.clone()) {
+                    return Err(Error::MultiplyDefined(arg.clone()))
+                }
+            }
+        } else {
+            let types = self.types.as_ref().unwrap(); // if types is None, and is_public = true, we should panic.
```

Comment:
Totally fine that this panics, but could you insert an error message here as well?

Reply:
Alternatively, you could make this type of function unrepresentable by removing `is_public` from `DefinedFunction` and instead create a type enumeration like this:

```
pub enum FunctionTypes {
   PublicFunction(DefinedFunction),
   PrivateFunction(DefinedFunction)
}
```

Then, you'd simply handle this as a `match` statement.

Reply:
Yep, I'll create a new type for this